### DISCUSSION TITLE
FIDO2 WebAuthn works, but not for Android App (NotAllowedError)

### DISCUSSION THREAD

**Discussion by melo0187** [dani-garcia/vaultwarden#3717] [Q&A]
I did setup vaultwarden on a Pi and allow access to it from the internet with a reverse proxy on my Synology NAS.
Everything seems to work fine, including websocket communication and two-step login with FIDO2 WebAuthn using my Solo 2 key.
Both the web vault and the Firefox extension successfully validate my login attempt using the security key as second login step.

However, when attempting to login with the official android app, the FIDO2 WebAuthn step can be initiated, but in the end fails.
It displays a dialog saying 
> NotAllowedError: The operation either timed out or was not allowed. See: https://www.w3.org/TR/webauthn-2/#sctn-privacy-considerations-client

Only thing configured on my reverse proxy is websocket support and as stated above everything works in the web vault and Firefox extension.
Any idea how I can dig deeper and find a solution?

Any help is much appreciated.
FIDO2 support was my main reason to attempt the migration from webdav synced keepass db to vaultwarden =)

Thanks in advance!

Support string:

### Your environment (Generated via diagnostics page)
* Vaultwarden version: v1.29.0
* Web-vault version: v2023.5.0
* OS/Arch: linux/arm
* Running within Docker: true (Base: Debian)
* Environment settings overridden: true
* Uses a reverse proxy: true
* IP Header check: true (X-Real-IP)
* Internet access: true
* Internet access via a proxy: false
* DNS Check: true
* Browser/Server Time Check: true
* Server/NTP Time Check: true
* Domain Configuration Check: true
* HTTPS Check: true
* Database type: MySQL
* Database version: 10.3.32-MariaDB
* Clients used: 
* Reverse proxy and version: 
* Other relevant information: 

### Config (Generated via diagnostics page)
<details><summary>Show Running Config</summary>

**Environment settings which are overridden:** DOMAIN, ADMIN_TOKEN, SMTP_HOST, SMTP_SECURITY, SMTP_PORT, SMTP_FROM, SMTP_USERNAME, SMTP_PASSWORD

```json
{
  "_duo_akey": null,
  "_enable_duo": true,
  "_enable_email_2fa": true,
  "_enable_smtp": true,
  "_enable_yubico": true,
  "_icon_service_csp": "",
  "_icon_service_url": "",
  "_ip_header_enabled": true,
  "_smtp_img_src": "cid:",
  "admin_ratelimit_max_burst": 3,
  "admin_ratelimit_seconds": 300,
  "admin_session_lifetime": 20,
  "admin_token": "***",
  "allowed_iframe_ancestors": "",
  "attachments_folder": "data/attachments",
  "authenticator_disable_time_drift": false,
  "data_folder": "data",
  "database_conn_init": "",
  "database_max_conns": 10,
  "database_timeout": 30,
  "database_url": "*****://**************************************************************",
  "db_connection_retries": 15,
  "disable_2fa_remember": false,
  "disable_admin_token": false,
  "disable_icon_download": false,
  "domain": "*****://******************",
  "domain_origin": "*****://******************",
  "domain_path": "",
  "domain_set": true,
  "duo_host": null,
  "duo_ikey": null,
  "duo_skey": null,
  "email_attempts_limit": 3,
  "email_expiration_time": 600,
  "email_token_size": 6,
  "emergency_access_allowed": true,
  "emergency_notification_reminder_schedule": "0 3 * * * *",
  "emergency_request_timeout_schedule": "0 7 * * * *",
  "enable_db_wal": true,
  "event_cleanup_schedule": "0 10 0 * * *",
  "events_days_retain": null,
  "extended_logging": true,
  "helo_name": null,
  "hibp_api_key": null,
  "icon_blacklist_non_global_ips": true,
  "icon_blacklist_regex": null,
  "icon_cache_folder": "data/icon_cache",
  "icon_cache_negttl": 259200,
  "icon_cache_ttl": 2592000,
  "icon_download_timeout": 10,
  "icon_redirect_code": 302,
  "icon_service": "internal",
  "incomplete_2fa_schedule": "30 * * * * *",
  "incomplete_2fa_time_limit": 3,
  "invitation_expiration_hours": 120,
  "invitation_org_name": "Vaultwarden",
  "invitations_allowed": true,
  "ip_header": "X-Real-IP",
  "job_poll_interval_ms": 30000,
  "log_file": null,
  "log_level": "Info",
  "log_timestamp_format": "%Y-%m-%d %H:%M:%S.%3f",
  "login_ratelimit_max_burst": 10,
  "login_ratelimit_seconds": 60,
  "org_attachment_limit": null,
  "org_creation_users": "",
  "org_events_enabled": false,
  "org_groups_enabled": false,
  "password_hints_allowed": true,
  "password_iterations": 600000,
  "push_enabled": true,
  "push_installation_id": "***",
  "push_installation_key": "***",
  "push_relay_uri": "https://push.bitwarden.com",
  "reload_templates": false,
  "require_device_email": false,
  "rsa_key_filename": "data/rsa_key",
  "send_purge_schedule": "0 5 * * * *",
  "sendmail_command": null,
  "sends_allowed": true,
  "sends_folder": "data/sends",
  "show_password_hint": false,
  "signups_allowed": false,
  "signups_domains_whitelist": "",
  "signups_verify": false,
  "signups_verify_resend_limit": 6,
  "signups_verify_resend_time": 3600,
  "smtp_accept_invalid_certs": false,
  "smtp_accept_invalid_hostnames": false,
  "smtp_auth_mechanism": null,
  "smtp_debug": false,
  "smtp_embed_images": true,
  "smtp_explicit_tls": null,
  "smtp_from": "************************",
  "smtp_from_name": "Vaultwarden",
  "smtp_host": "*********",
  "smtp_password": "***",
  "smtp_port": 587,
  "smtp_security": "starttls",
  "smtp_ssl": null,
  "smtp_timeout": 15,
  "smtp_username": "************************",
  "templates_folder": "data/templates",
  "tmp_folder": "data/tmp",
  "trash_auto_delete_days": null,
  "trash_purge_schedule": "0 5 0 * * *",
  "use_sendmail": false,
  "use_syslog": false,
  "user_attachment_limit": null,
  "web_vault_enabled": true,
  "web_vault_folder": "web-vault/",
  "websocket_address": "0.0.0.0",
  "websocket_enabled": false,
  "websocket_port": 3012,
  "yubico_client_id": null,
  "yubico_secret_key": null,
  "yubico_server": null
}
```
</details>

**Comment by zacknewman** [comment]
Odds are this is a client issue and not a Vaultwarden issue. Use your favorite search engine to discover a bunch of problems related to WebAuthn and the Bitwarden Android app. I know I cannot use the Android app. I don't mind using the web-vault or the Firefox extension though. Also note that the only desktop applications that support WebAuthn are the Windows 10 and Windows 11 versions (e.g., if you use Linux, you're SOL). No version of the CLI app supports WebAuthn.

You can try using the official Bitwarden server and see if the Android app works if you are somehow confident this is in fact a Vaultwarden problem. I would be pretty surprised if it is though.

**Comment by zacknewman** [comment]
Ah, so I finally tried again and figured out why WebAuthn does not work for me. In my case it is due to my older phone running Android 9. This causes an issue with the more secure cipher suites that are required. I should have known this was the issue as that is the same reason I had to support TLS 1.2 on my SMTPS and IMAPS servers. Anyway, I tried on my wife's phone; and I had no issue getting my YubiKey 5Ci with firmware 5.2.6 to work. The version of Android on her phone is Android 11. Unfortunately I don't have a Solo 2 to test

**Reply by melo0187** [reply]
@zacknewman thank your for your quick reply and looking into this so thoroughly.

My experience now seems explainable based on the information you provide. My Nexus 5X (yeah I know...) is stuck with Android 8. I also tested with my wife's phone and found the same issue. Her phone is a Sony Xperia XZ2 compact and while I'm not 100% sure which Android version is currently installed, it seems the newest you can get on that one is Android 10.

Guess I didn't do enough research when buying Solo 2 with NFC hoping it would work on my 8 years old phone... :rofl: 

Anyway, I'm happy to use bitwarden's Android app with TOTP as MFA in the meantime.

Thanks again for your quick response and for vaultwarden in general. Have been using it only for a day, but everything is relatively easy to setup, packs a ton of nice features, just works and is also amazingly fast on a Pi 2 (I guess Rust can really shine here)! Last migration steps are moving over attachments from kdbx to my new vault and setup a proper backup and I'm a happy vaultwarden user for the time being =)

**Reply by zacknewman** [reply]
> @zacknewman thank your for your quick reply and looking into this so thoroughly.

Lol, you're welcome; but there was nothing "thorough" about this. I spent less than 5 minutes trying the Android app on the phones that belong to my wife and me.

> Guess I didn't do enough research when buying Solo 2 with NFC hoping it would work on my 8 years old phone... rofl

I don't think it's fair to blame the Solo 2. You successfully use it for WebAuthn with the Firefox extension and the web-vault. You can likely use it for other applications like Google's Advanced Protection program. If there is something to blame, it's your phone or Android and their planned obsolescence.

> Anyway, I'm happy to use bitwarden's Android app with TOTP as MFA in the meantime.

I personally prefer the increased security WebAuthn provides over TOTP more than my desire to use the Android app, so I simply restrict my use to the Firefox extension and web-vault. Both of those work on all of my devices including my phone. Obviously if you insist on using the Android app, then you either have to use a less secure form of 2FA (e.g., TOTP) or get a new phone.

Last, unrelated to this but regarding the following:

> Only thing configured on my reverse proxy is websocket support and as stated above everything works in the web vault and Firefox extension.

You don't need a reverse proxy server for Websockets anymore as Rocket now supports that. Unless you have [other reasons](https://github.com/dani-garcia/vaultwarden/discussions/3718#discussioncomment-6513400) for a reverse proxy server, you can remove the middleman.

**Reply by melo0187** [reply]
> I spent less than 5 minutes

I'm just some unknown guy using free software, so even those few minutes don't come naturally and are much appreciated :-)

> I don't think it's fair to blame the Solo 2

I guess my statement came out wrong. I don't blame Solo, but only myself for not properly researching what it would take to use webauthn on mobile. My expectation for this to just work on my phone was misinformed and is in the end unmet because of my phone being so old. Still very happy with the Solo 2 all in all.

My desire to use the app is caused by its auto-fill capabilities.

The fact I use a reverse proxy is simply because I run multiple services and prefer to put them all behind a reverse proxy in conjunction with the use of subdomains, which allows me to only forward port 443 from my router to the reverse proxy where I can manage everything else.

**Reply by zacknewman** [reply]
> My desire to use the app is caused by its auto-fill capabilities.

The Firefox extension has that as well including on Android.

**Reply by zacknewman** [reply]
Hold the press! My issue was likely unrelated to yours. I force TLS 1.3 to be used by Rocket, so my solution for the Android app is similar to my issue with the Android mail app when interacting with my SMTPS and IMAPS servers: allow TLS 1.2.

When I forced TLS 1.3, the app complained about a "transient error" which is not the error you're getting. Once I enabled TLS 1.2 on Vaultwarden, this error went away; however I did get a timeout error a couple times. The app is _very_ finicky and required me to repeatedly try and tap my key a bunch of times until it finally worked. This behavior [is consistent with other people's experiences](https://www.reddit.com/r/Bitwarden/comments/pw6v29/yubikey_fido2_webauthn_on_android_it_works_but/); as a result, your Solo 2 may very well work, but it can be a pain. This is not a Vaultwarden issue though but a Bitwarden one.

Do note that if/when you do finally get it to work, you can lock the vault instead of log out which will allow you to simply type in your master password alone. Of course that defeats the purpose of 2FA, so I personally wouldn't recommend that.

**Reply by melo0187** [reply]
:thinking: makes me wonder for several reasons.

1. Android 8.1 apparently [supports TLS 1.3](https://www.ssllabs.com/ssltest/viewClient.html?name=Android&version=8.1&key=157), but has it enabled by default [only since version 10](https://developer.android.com/about/versions/10/behavior-changes-all#tls-1.3).
2. I can't find any option to configure enabled TLS versions in vaultwarden. Can you describe how you did this?

**Reply by zacknewman** [reply]
As stated, none of this likely matters to you since you were not getting a "transient error". I have not rooted my phone or gone to any steps to enable TLS 1.3 and its cipher suites, so I cannot use TLS 1.3-_only_ services unless the client application has integrated such functionality directly (e.g., Firefox).

There is no reason to know how to enable specific TLS cipher suites in Rocket since you don't use TLS in Vaultwarden. As you said, you rely on a reverse proxy server which very likely forwards traffic to Vaultwarden via HTTP _not_ HTTPS. If you haven't mucked around your reserve proxy server's settings vis-à-vis TLS versions or cipher suites, then it's _very_ unlikely TLS 1.2 is disabled since a subset of TLS 1.2 ciphers are still considered secure. If you do bizarrely forward HTTPS traffic to Vaultwarden, then it still won't matter since Rocket enables the secure ciphers of TLS 1.2 by default.

After messing around a little more, my issue was that I had the OpenKeychain app installed on my phone. Anytime I plugged in my YubiKey, I would be prompted with "Allow Google Play Services to access Yubikey OTP+FIDO+CCID?"; and this would cause issues with WebAuthn in the Bitwarden app. I would have to try a bunch of times reconnecting my YubiKey and trying to tap the gold contacts quickly enough so that prompt wouldn't show up to get it to work. Once I uninstalled that app, I haven't gotten that prompt the past few times. If you read the supplied link showing other people's issues with WebAuthn and the Bitwarden Android app, you will see other people complaining about that prompt too.

**Reply by zacknewman** [reply]
TL;DR

1. HTTPS server didn't have compatible TLS ciphers with the Android 9 Bitwarden app which caused a "transient error" to be displayed. Many applications (e.g., Android Bitwarden app) use `libssl` instead of bundling their own SSL/TLS library like many web browsers do. The solution was to enable at least one compatible TLS cipher on the HTTPS server. In my case that simply meant removing the Rocket configuration I had that was forcing TLS 1.3 ciphers since by default Rocket contains at least one compatible cipher with Android 9's `libssl`.
2. Applications installed on my phone—in my case OpenKeychain—caused a prompt anytime I plugged in my YubiKey. This prompt caused "timeout" issues with WebAuthn in the Bitwarden app. If you were "lucky", you could get WebAuthn to work _before_ this prompt; but often times this prompt would "win" and disrupt WebAuthn forcing you to try again. In my case I uninstalled the problematic app(s), and I haven't received the prompt and thus "timeout" errors.
